### DISCUSSION TITLE
Add \_Von to bibmark when `\nonumcitations` is set

### DISCUSSION THREAD

**Discussion by pabloabur** [olsak/OpTeX#167] [Q&A] [answered]
Hi,

Is there a recommended way of adding von/van/da/etc before lastnames in the output of `\cite`?

I tried changing `\_authorname` (to get \_Lastname and \_Von only) and `\_createbibmark` but I am not sure whether there are bad side effects from this. Also, there is a space before the first author name that I cannot get rid of. In the example below, the output looks like this:

```
Citation [ Lastname1, Lastname2 and van Lastname3, 2024].
          ^

```
I tried to create a temporary variable inside `\_createbibmark` but that leads to an error. I was thinking that I could do something like `\def\temp{#3}\ifx\temp\empty\else#3\fi #2,` but that does not work for some reason. Has anybody done this before?

MWE
test.tex:
```
\fontfam[LMfonts]

\nonumcitations
\bibtexhook{
    \def\_createbibmark #1;#2,#3;#4,#5;#6,#7;#8,#9\_fin{%
        \ifx ^#8^%
            \ifx ^#6^%
                \ifx ^#4^%
                    #3 #2,
                \else
                    #3 #2 and #5 #4,
                \fi
            \else #3 #2, #5 #4 and #7 #6,
            \fi
        \else #3 #2 et al.,
        \fi
        \space #1% 
    }
    \_def\_authorname{%
        \_ifx\_authlist\_undefined \_edef\_authlist{\_Lastname,\_Von}%
            \_else \_edef\_authlist{\_authlist;\_Lastname,\_Von}\_fi
            \_ifnum\_NameCount>0\_namecount\_relax\_else \_commonname \_fi
            \_ifnum\_NameCount=0\_namecount\_relax \_maybeetal \_fi
    }

}

Citation \cite[ref].

\bigskip
\usebib/s (iso690) bibsample
\bye
```
biblist.bib:
```
@ARTICLE{ref,
    title="title",
    author="Lastname1, Firtname1 and Lastname2, Firstname2 and van Lastname3, Firtname3",
    year="2024"
    }
```

**Comment by olsak** [comment]
The `\_createbibmark` macro must be expandable because its output is expanded to the .ref file (you can look into this file). If you want to have `\def` here, you have to prefixed it by `\immediateassignment`. But it isn't best idea. I will try something else and post it here yesterday.  The space before Lastname1 is due to `#3 #2` where `#3` is empty "von" and `#2` is Lastname1. The space between `#3 #2` is this problematic space. I will try to suggest a slightly different solution. The question: do you want to print maximal 3 authors in your \cite or possibly more?

**Reply by pabloabur** [reply]
Thank you, I appreciate it. Yes, I would like 3 to be the maximum. More than that should become Lastname1 et al. (year).

**Comment by olsak** [comment, marked as answer]
You can re-define `\_authorname` if you plan to use only `bib-iso690`, because the `\_authorname` macro in `bib-simple` is defined differently. Then, you can use following macros:

```
\fontfam[LMfonts]

\nonumcitations
\bibtexhook{
    \def\_createbibmark #1;#2,#3;#4,#5;#6,#7;#8,#9\_fin{%
       \ifx ^#4^% single author
          \printvon{#3}#2,
       \else \ifx ^#6^% two authors
          \printvon{#3}#2 and \printvon{#5}#4,
       \else \ifx ^#8^% three authors
          \printvon{#3}#2, \printvon{#5}#4 and \printvon{#7}#6,
       \else          % more authors
          \printvon{#3}#2, \printvon{#5}#4, \printvon{#6}#7 et al.,
       \fi\fi\fi
    #1% year
    }
    \def\printvon#1{\ifx^#1^\else #1 \fi} % print von + space

    \_def\_authorname{%
        \_ifx\_authlist\_undefined \_edef\_authlist{\_Lastname,\_Von}%
            \_else \_edef\_authlist{\_authlist;\_Lastname,\_Von}\_fi
            \_ifnum\_NameCount>0\_namecount\_relax\_else \_commonname \_fi
            \_ifnum\_NameCount=0\_namecount\_relax \_maybeetal \_fi
    }

}

Citation \cite[ref].

\bigskip
\usebib/s (iso690) bibsample
\bye
```
Second variant is to keep `\_authorname` and `\_authlist` without changes and define only `\_createbibmark`. This definition is more tricky:

``` 
\fontfam[LMfonts]

\nonumcitations
\bibtexhook{
%    \def\_createbibmark #1;#2\_fin{\createbibmarkA #2\_fin #1} % from previous experiments
    \def\_createbibmark #1;#2,#3,#4,#5;#6,#7;#8,#9\_fin{%
       \printvon{#4}#2%
       \ifx^#6^, % single name
       \else
          \ifx^#8^ and \createbibmarkB #6,#7;,   % two names
          \else    , \createbibmarkB #6,#7;%
                     \createbibmarkC #8,#9\_fin  % three or more names  
          \fi
       \fi
       #1% year
    }
    \def\createbibmarkB #1,#2,#3,#4;{\printvon{#3}#1}
    \def\createbibmarkC #1,#2,#3,#4;#5,#6\_fin {%
       \ifx ^#5^ and \printvon{#3}#1,        % three names
       \else     ,   \printvon{#3}#1 et al., % more names
       \fi
    }
    \def\printvon#1{\ifx^#1^\else #1 \fi} % print von + space
}

Citation \cite[ref].

\bigskip
\usebib/s (iso690) bibsample
\bye
```

**Reply by pabloabur** [reply]
Thanks, both of these solutions work for me.

Could you please briefly explain `\def\_createbibmark #1;#2\_fin{\createbibmarkA #2\_fin #1}` though? As far as I understand, `\_fin` is like a placeholder for other arguments, so you can process more than 9 arguments, right? But I don't really get this line, especially since `\createbibmarkA` is not defined anywhere

**Reply by olsak** [reply]
Oops, I forgot to delete the line `\def\_createbibmark #1;#2\_fin{\createbibmarkA #2\_fin #1}` from my code. It was here from my previous experiments. The `\_createbibmark` is re-defined immediately at the next line. My previous experiment included `\_def\createbibmarkA #1,#2;#3,#4;#5,#6;#7,#8\_fin{...}` which prints first author `#1,#2` and maybe second author `#3,#4` (if not empty) etc. The `#8` includes the data left after reading the first three authors finalized by `\_fin`. The "year" (`#1` in `\_createbibmark`) is printed after the `\createbibmarkA` is finished.

**Reply by pabloabur** [reply]
Great, I think I get it now.
